### DISCUSSION TITLE
Razer RGB control when Windows is locked

### DISCUSSION THREAD

**Discussion by Blackbird714** [devcompl/TTController#114] [Q&A] [answered]
Hello!  First wanted to thank you for the great software!

Just finished my first build with ThermalTake cooling.  Everything is installed in a Lian Li O11 Dynamic Razer Edition with some other Razer peripherals.  While the TT hardware is great, the "TT RGB Plus" software, not so much.  Was seeing really flaky behavior -- when I contacted support they wanted to RMA the fan controllers.  But the way it was behaving, I was pretty convinced it was software.

Ran across your project -- and I thought I would be losing features, specifically Razer Chroma Connect, but to my surprise, everything just worked.  Well, some assembly required to get the config.json working... but aside from that, your software seems to work more reliably than TT's solution.

Now my question; I've observed the TTController service startup and start synchronizing colors with my system without intervention.  TT's software definitely did not do that.  Kudos on making that work!

However after logging in, if I plan to leave my desk, I use Win+L to lock my screen.  When Windows is locked, the "TT RGB Plus" software seemed to just get suspended in place -- the fans and AIO would just freeze at whatever color they were when the lock event happened.  TTController seems to lose contact with the fan controllers entirely, so they revert to the default rainbow pattern.

Neither behavior is ideal -- would love to have it continue to sync with the rest of the system, which Razer seems to control just fine (using the "Spectrum Cycling" quick effect) even when the system is locked.

I have attached my config.json in case you have any suggestions.  
[config.json.txt](https://github.com/MoshiMoshi0/TTController/files/5756107/config.json.txt)

ThermalTake products in use:
Floe Riing RGB 360 TT Premium Edition
2 x Pure Plus 12 RGB Radiator Fan TT Premium Edition (3-Fan Pack)

**Comment by MoshiMoshi0** [comment]
That must be a bug, it works just fine for me when windows locks itself, I did not test manual Win+L tho.
I wonder if your "Suspend" profile is causing that, maybe it thinks it is suspending when you lock your PC. Try removing that profile.

I will investigate further later.

**Reply by Blackbird714** [reply]
Actually my initial testing did not include "StateType": "Suspend" at all; I added it hoping that might alter the behavior that I describe.  But it did not change anything.

I just setup a blank screen saver that locks my screen; curious to see if that behaves differently.  Will post results from that test.

**Reply by MoshiMoshi0** [reply]
I tested locking with Win+L. I also have a blank screensaver so that would not be it.

Can you post the log file when you lock the PC and the fans freeze?

**Reply by Blackbird714** [reply]
See attached debug log; I pressed Win+L a little after 10:04:22.5860.

I unlocked/logged back in around 10:04:37.5907.

I'm not seeing any log events that correlate with the lock, and it looks like fan and temp data continues to be logged.

[debug.txt](https://github.com/MoshiMoshi0/TTController/files/5757561/debug.txt)

**Comment by MoshiMoshi0** [comment]
Edit `TTController.Service.exe.Config` file and change line:
`      <logger name="*" minLevel="Info" writeTo="file" enabled="true" />`
to
`      <logger name="*" minLevel="Trace" writeTo="file" enabled="true" />`

Rerun until the problem happends, then post `Logs\current.log` file.

I think this is an issue with razer software, it just stops sending data when you ~~go to sleep~~ lock the PC.
Quick way to test would be to change one of `"RazerConnectEffect"` to `"AuroraEffect"`

**Reply by Blackbird714** [reply]
The current.log is attached.  Lock happened around 10:25:24, unlock around 10:25:34.

[current.log](https://github.com/MoshiMoshi0/TTController/files/5757607/current.log)

**Reply by Blackbird714** [reply]
I switched from "RazerConnectEffect" to "AuroraEffect" as you suggested -- and locking Windows did not change the lighting behavior.  (The TT RGB did not revert to spinning rainbow/wave effect.)

Pending your assessment of the logs -- this observation seems to confirm that Razer Chroma SDK Server/Service stops sending useful data.  Yet the RGB built into the case and the peripherals continue to receive data just just fine.

So I guess this means this is a question that I would have to bring to Razer regarding their Chroma Connect integration?

**Reply by MoshiMoshi0** [reply]
I dont understand your log. It never sends any colors, its all black.

What colors are you sending with razer synapse. 
What synapse version are you using.
You should use [IgnoreColorCache](https://moshimoshi0.github.io/TTController/config/port#ignorecolorcache) option on all Riing Quad controller ports.
You should be setting [ColorModifiers](https://moshimoshi0.github.io/TTController/config/port#colormodifiers) with [LerpLedColorModifier](https://moshimoshi0.github.io/TTController/plugins/modifiers/lerpledcolormodifier) when using razer effect since it outputs only 4 colors.

I tested the effect on my machine and it works just fine when I lock windows.

**Reply by Blackbird714** [reply]
Razer is set to the "Spectrum Cycling" quick effect.  The version of Razer Synapse is showing that it is the latest available:  3.5.1215.121019

When I played around with the IgnoreColorCache parameter, RGB stopped working entirely.  Can you give me an example of what the PortConfigs would look like?

```json
  "PortConfigs": [
    {
      "Ports": [
        [9802, 8102, 5]
      ],
      "Config": {
        "DeviceType": "FloeRiing"
      }
    }
  ],
```

**Reply by MoshiMoshi0** [reply]
```json
    {
      "Ports": [
        [9802, 8501, 5]
      ],
      "Config": {
        "DeviceType": "FloeRiing",
        "ColorModifiers": [
          {
            "Type": "LerpLedColorModifier"
          }
        ]
      }
    },
    {
      "Ports": [
        [9802, 8801, 1],
        [9802, 8801, 2],
        [9802, 8801, 3],
        [9802, 8802, 1],
        [9802, 8802, 2],
        [9802, 8802, 3]
      ],
      "Config": {
        "IgnoreColorCache": true,
        "ColorModifiers": [
          {
            "Type": "LerpLedColorModifier"
          }
        ]
      }
    }
```

**Reply by Blackbird714** [reply]
Ok that's the nudge in the right direction that I needed.  Thanks!

Initially though, I thought it was messing up like my previous experiments -- when I stopped the service and restarted, the RGB went out and would no longer respond.  This time I decided to try a clean boot -- and things started to behave correctly again after Windows restarted.

Now when I lock with Win+L, the colors freeze in time.  That is exactly the same behavior that I was seeing with RGBRiingPlus -- not ideal, but much less obnoxious than the default rainbow wave.

I see a couple of posts in the Razer community about RGB not working as expected on the Windows lock screen.  I will see if I can dig anything else up, and post it here if I find anything useful.

**Comment by Blackbird714** [comment, marked as answer]
Just closing the loop; Moshi, thank you for helping me get my config.json updated to get the RGB behavior to work at least as well as it was with TT's software.  Fixed file is attached.

[config.json.txt](https://github.com/MoshiMoshi0/TTController/files/5759378/config.json.txt)

If I ever come up with an answer /wrt to behavior when Windows is locked, I will post an update on this thread.  Here is the topic I created in the Razer Support forum:

https://insider.razer.com/index.php?threads/razer-chroma-devices-and-windows-lock.68225

**Reply by MoshiMoshi0** [reply]
Just to be sure, when using thermaltake "TT RGB PLUS" software, was the behavior the same?

**Reply by Blackbird714** [reply]
Yes, "TT RGB PLUS" behaved the same /wrt to Chroma Connect -- when I lock Windows, most of the Chroma devices continue "Spectrum Cycling" in sync, while the TT fans (and my G.Skill memory) stay at whatever color they are at the time the system is locked.

**Reply by Blackbird714** [reply]
I received a response from Razer indicating that this is a "limitation" not a "bug" -- I suspect that means that this is something they have no plan to address.

I do see messages like this being logged by their SDK service when the PC is locked/unlocked:

> [2021-01-02 11:23:40][Info][ChromaSDKService] Session event received
> [2021-01-02 11:23:52][Info][ChromaSDKService] Session event received

The problem is I don't know enough about how this interprocess communication works, and how a Windows lock affects it -- haven't found any clear documentation on the topic.  But if TTController can see that this happened, maybe we could set a specific color profile -- similar to StateType (Shutdown, Boot, Suspend)?

**Reply by MoshiMoshi0** [reply]
Like I said for me it was working when locking, but i only have razer sdk without synapse. Not sure how a windows lock can be a limitation. Ive done some work on razer internals and they have a couple of those weird "limitations", mainly forcing you to buy/use their stuff.

I'll see what dumb stuff they are doing. I will for sure not add anything special to handle razer into the main service, if anything it will be in razer plugin. But also im not sure if I should be adding some internal razer stuff to a public repository.
